servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/BaseHostSelector.java

```diff
@@ -27,17 +30,53 @@ abstract class BaseHostSelector<ResolvedAddress, C extends LoadBalancedConnectio
         implements HostSelector<ResolvedAddress, C> {
 
     private final String targetResource;
-    BaseHostSelector(final String targetResource) {
+    private final List<Host<ResolvedAddress, C>> hosts;
+    BaseHostSelector(final List<Host<ResolvedAddress, C>> hosts, final String targetResource) {
+        this.hosts = hosts;
         this.targetResource = requireNonNull(targetResource, "targetResource");
     }
 
+    protected abstract Single<C> selectConnection0(Predicate<C> selector, @Nullable ContextMap context,
+                                         boolean forceNewConnectionAndReserve);
+
+    @Override
+    public final Single<C> selectConnection(Predicate<C> selector, @Nullable ContextMap context,
+                                      boolean forceNewConnectionAndReserve) {
+        return hosts.isEmpty() ? noHostsFailure() : selectConnection0(selector, context, forceNewConnectionAndReserve);
+    }
+
+    @Override
+    public final boolean isHealthy() {
+        // TODO: in the future we may want to make this more of a "are at least X hosts available" question
+        //  so that we can compose a group of selectors into a priority set.
+        return !allUnhealthy(hosts);
+    }
+
     protected final String getTargetResource() {
         return targetResource;
     }
 
-    protected final Single<C> noActiveHosts(List<Host<ResolvedAddress, C>> usedHosts) {
+    protected final Single<C> noActiveHostsFailure(List<Host<ResolvedAddress, C>> usedHosts) {
         return failed(Exceptions.StacklessNoActiveHostException.newInstance("Failed to pick an active host for " +
                         getTargetResource() + ". Either all are busy, expired, or unhealthy: " + usedHosts,
                 this.getClass(), "selectConnection(...)"));
     }
+
+    private Single<C> noHostsFailure() {
+        return failed(Exceptions.StacklessNoAvailableHostException.newInstance(
+                "No hosts are available to connect for " + targetResource + ".",
+                this.getClass(), "selectConnection(...)"));
+    }
+
+    private static <ResolvedAddress, C extends LoadBalancedConnection> boolean allUnhealthy(
```

Comment:
I see what you're suggesting but I'm going to tweak it to keep the logic: for `isHealthy` what we're defining it internally as is "are there _any_ healthy?", not "are all healthy?" Of note, in the future we may also change it to "are at least X healthy?"

Reply:
I ended up going yet another way: I decided to use `isUnhealthy()` as the abstract method name because it corresponds to the method on `Host`. Maybe we can change them both, but I decided not to do that right now.